packages/tree-sitter-port/src/prepareNodeTypes.ts

```diff
@@ -0,0 +1,297 @@
+    types: Array<{
+      type: string
+      named: boolean
+    }>
+  }
+
+  subtypes?: Array<{
+    type: string
+    named: boolean
+  }>
+}>
+
+type FieldsMeta = {
+  type: string
+  singleFieldNames: string[]
+  nodeTypeToMultipleFieldName: Record<string, string>
+  multipleOrChildrenFieldNames: string[]
+}
+
+export type ConflictResolver = {
+  nodeType: string
+  fields: [string, string]
+  mergeToField: string
+}
+
+export const prepareNodeTypes = ({
+  nodeTypesToIgnore,
+  nodeTypesFilePath,
+  conflictResolvers = [],
+  log,
+}: {
+  nodeTypesToIgnore: string[]
+  nodeTypesFilePath: string
+  conflictResolvers?: Array<ConflictResolver>
+  log: (msg: string) => void
+}) => {
+  const nodeTypes: NodeTypes = JSON.parse(
+    fs.readFileSync(nodeTypesFilePath).toString(),
+  )
+
+  const types: Record<
+    string,
+    {
+      named: boolean
+      type: string
+      fieldsMeta: null | Record<
+        string,
+        {
+          multipleOrChildren: boolean
+          nodeTypes: string[]
+        }
+      >
+    }
+  > = {}
+
+  /**
+   * Not named types are usually some keywords or language constants.
+   * We don't care about them unless they two or more of not named types can occur in the same place in tree.
+   * We capture that when processing nested node types and filter out all Not named except those relevant.
+   */
+  const notNamedNodeTypesToInclude: string[] = []
+
+  const nodeTypesWithSubTypes = nodeTypes.filter(({ subtypes }) => subtypes)
+
+  const subTypesMap = Object.fromEntries(
+    nodeTypesWithSubTypes.map(({ type, subtypes }) => [
+      type,
+      subtypes?.map(({ type }) => type),
+    ]),
+  )
+
+  /**
+   * Some subtypes are nested, hence recursion
+   *  */
+  const extrapolateBySubtype = (nodeType: string): string[] =>
+    subTypesMap[nodeType]
+      ? subTypesMap[nodeType]?.map(extrapolateBySubtype)?.flat() ?? []
+      : [nodeType]
+
+  nodeTypes.forEach((nodeType) => {
+    const fieldsRaw = nodeType.fields ?? {}
+
+    if (nodeType.children) {
+      fieldsRaw.children = nodeType.children
+    }
+
+    const fieldsMeta = fieldsRaw
+      ? Object.fromEntries(
+          Object.entries(fieldsRaw).map(([fieldName, value]) => {
+            const notNamedNodeTypesInField = value.types
+              .filter(({ named }) => !named)
+              .map(({ type }) => type)
+            notNamedNodeTypesToInclude.push(...notNamedNodeTypesInField)
+
+            return [
+              fieldName,
+              {
+                multipleOrChildren: value.multiple || fieldName === 'children',
+                /*
+                 *  Some types are just aliases for types group, eg. "expression" can be "boolean_operator" "primary_expression" and others
+                 * Some subtypes are nested so we do recursion, hence we have to flat.
+                 */
+                nodeTypes: value.types
+                  .map(({ type }) => extrapolateBySubtype(type))
+                  .flat(10),
```

Comment:
magic number ?

Reply:
Yeah, it's just a reasonable depth value. I will give it a name

Reply:
Turned out it can be just 1

Reply:
Recursion magic 🧙